fixes #21694 - Add mtu to subnet

bagasse opened the pull request:
* Add MTU to subnet
* API v2 modification in a separate commit (I don't know if you want to modify the API or not)

Commit:
fixes #21694 - Add MTU to subnet

Commit:
fixes #21694 - Add MTU to subnet and interface API

timogoebel commented:
@bagasse: Code looks good. Thanks for this, this is a great addition.
Do you think there might be cases when a user wants to overwrite the subnet's mtu on a per-interface basis?

bagasse commented:
Differentiate MTU size between IPv4 and IPv6 and adding some tests around it.

@timogoebel, I don't see any reason to apply a different MTU on a network.

sean797 commented on app/models/subnet/ipv6.rb:
```diff
@@ -9,6 +9,7 @@ class Subnet::Ipv6 < Subnet
   has_many :primary_hosts, :through => :primary_interfaces, :source => :host
 
   validate :validate_eui64_prefix_length, :if => Proc.new { |subnet| subnet.ipam == IPAM::MODES[:eui64]}
+  validates :mtu, :numericality => {:only_integer => true, :greater_than_or_equal_to => 1280, :less_than_or_equal_to => 65536}, :allow_nil => true
```
`less_than_or_equal_to` should be 4,294,967,295 as per rfc 2675

bagasse replied:
I was on the rfc8200 that don't specify maximum MTU, so I reported the IPv4 one. Your RFC talk about Jumbo datagram. I nerver saw that :). Fixed.

sean797 commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,23 @@ def vlanid
+      # user's database. There is no way of determining the real mtu, so we
+      # pick the v4 one unless it turns out to be blank.
+
+      #return self.tag unless self.tag.blank?
```
Can you remove this line?

bagasse replied:
Done.

sean797 reviewed (changes requested):
Overall looks good thanks @bagasse 👍

@timogoebel I think there is a case for per interface, though i think we could do that in another PR if someone wants it.

Note that this sort of thing can currently be done using subnet Parameters, though I do think MTU does deserve its own field, but we may need to draw the line somewhere in future.

sean797 commented on app/views/subnets/_fields.html.erb:
```diff
@@ -22,6 +22,7 @@
   <%= text_f f, :to,   :help_inline => _("Optional: Ending IP Address for IP auto suggestion") %>
 </div>
 <%= number_f f, :vlanid, :min => 0, :max => 4095, :help_inline => _("Optional: VLAN ID for this subnet") %>
+<%= number_f f, :mtu, :min => 1, :max => 65536, :help_inline => _("Optional: MTU for this subnet") %>
```
Same thing here ;)

ares commented on app/controllers/api/v2/subnets_controller.rb:
```diff
@@ -43,6 +43,7 @@ def show
           param :from, String, :desc => N_("Starting IP Address for IP auto suggestion")
           param :to, String, :desc => N_("Ending IP Address for IP auto suggestion")
           param :vlanid, String, :desc => N_("VLAN ID for this subnet")
+          param :mtu, String, :desc => N_("MTU for this subnet")
```
Should this perhaps be a `:number` instead of a string?

bagasse replied:
As VLANID was a String, just copied it. Changed to Integer.

ares commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,22 @@ def vlanid
+      # In case the v4 and v6 subnet are both present, they should have the same
+      # mtu. If they have a different mtu, this is probably an error in the
+      # user's database. There is no way of determining the real mtu, so we
+      # pick the v4 one unless it turns out to be blank.
```
It would be great to display a warning to the user if they chose two subnets with different mtu for single interface but I don't insist.

bagasse replied:
Yes, and it would be nice to have a warning in case of compute resource network with different MTU too. I can manage the oVirt part (and libvirt if needed), but I don't have any openstack/vmware/aws/azure compute resource to test other implementations.

It would be nice for me if we can keep this warning things out of this PR, to let it as simple as possible :)

ares replied:
I didn't mean to ask compute resource about the setting. What I meant is that if you choose in Foreman subnet for ipv4 and ipv6 for a single interface, we could check what MTU is set for them and display warning if these two subnets have different value. It would only be based on Foreman data.

bagasse replied:
Yes, sorry maybe my sentence was not clear. I was trying to say that we can check that both subnet's mtu are identical and in the case we are provisioning an VM, we can add another check with the compute resource network mtu too, as when we configure the nic for a VM, we already get the compute resource network informations.

ares replied:
ok, so I suppose the outcome is we'll keep it for later

ares commented on app/models/subnet.rb:
```diff
@@ -16,7 +16,7 @@ class Subnet < ApplicationRecord
 
   attr_exportable :name, :network, :mask, :gateway, :dns_primary, :dns_secondary, :from, :to, :boot_mode,
-    :ipam, :vlanid, :network_type, :description
+    :ipam, :vlanid, :mtu, :network_type, :description
```
:+1:

ares reviewed (commented):
Did you consider ntu being a required attribute defaulting to 1500 (or whatever is the best default value)? I think it would make it a bit more usable since users wouldn't have to add too many if conditions when using the value, e.g. in templates or puppet modules.

bagasse commented:
@ares, if it make sense to you, I can set 1500 as default value for MTU. I didn't do that to be sure to not break any installed instance if some users. I pushed the change with default value to 1500 and I will update templates to reflect that. If someone have an argument to not do things like this, I will rollbakck to the 'no value' default.

sean797 commented:
I personally think we shouldn't have a default, the OS should have a default, making a default just gives us more code to maintain with no real benefit IMO. Unless @ares or someone else has a good reason for a default?

timogoebel commented:
My only concern with this (that's why I was asking for a per interface value) is the following: How do we handle fact imports for this? We import all other data for a NIC from facts. But not the MTU. I think we should be consistent here. What do you guys think? Can we just add that later or do we need it now?

timogoebel commented on db/migrate/20171119094913_add_mtu_to_subnet.rb:
```diff
@@ -0,0 +1,5 @@
+class AddMtuToSubnet < ActiveRecord::Migration
+  def change
+    add_column :subnets, :mtu, :integer
```
This does not store `4294967295` on mysql, as passing no limit to the migration makes rails create a `integer` column on mysql. The maximum value is `2147483647`. By setting something between `limit: 5` and `limit: 8` causes rails to create a bigint column with a maximum value of `9223372036854775807`.

bagasse replied:
Done.

ares commented:
@sean797 I don't insist on default, though in all templates and puppet manifest we'd need to do "if mtu set then do something with it else ignore". I think changing mtu is quite rare so for most people the default 1500 would do. The default should add zero code, it's only about setting the default value in SQL DB for the column, every new instance would then be initialized with 1500. The only code required is `validates :mtu, :presence => true` in subnet model.

@timogoebel I think we store additional interface attributes serialized in attrs. I think it would be a good idea to have it as a separate attribute but I think it does not have to be part of this PR. Perhaps we could have `interface.mtu` method delegated to it's subnet so later it's easy to add new field without templates change?

bagasse commented:
@ares, @timogoebel the `interface.mtu` is already there and redirect to `subnet.mtu`.

bagasse commented:
[test katello]

bagasse commented:
[test katello]

sean797 commented on app/views/subnets/_fields.html.erb:
```diff
@@ -22,6 +22,7 @@
   <%= text_f f, :to,   :help_inline => _("Optional: Ending IP Address for IP auto suggestion") %>
 </div>
 <%= number_f f, :vlanid, :min => 0, :max => 4095, :help_inline => _("Optional: VLAN ID for this subnet") %>
+<%= number_f f, :mtu, :min => 1, :max => 4294967295, :help_inline => _("Optional: MTU for this subnet") %>
```
`:min => 68` ?

bagasse replied:
done.

sean797 reviewed (changes requested):
Thanks @ares we can set a default if you like. It would mean that users using some funny drivers might have to override it, but i think that's fairly rare.

I'm not sure how setting MTU on bridge, and bonded interfaces will be effected, as  believe bonded interfaces only need it set on the bond and not the salves, but I guess that is a template issue.

bagasse commented:
Re-added default MTU to 1500 and rebased. MTU for bond slave don't matter, the MTU set on the bonding iface will be used. You set the MTU on a bridge like any other interface.

sean797 commented on db/migrate/20171119094913_add_mtu_to_subnet.rb:
```diff
@@ -0,0 +1,5 @@
+class AddMtuToSubnet < ActiveRecord::Migration
+  def change
+    add_column :subnets, :mtu, :integer, :default => 1500, :null => false, :limit => 8
```
Should it be `:limit => 10` to allow for `4294967295` ?

bagasse replied:
Limit here is the number of bytes used to store the integer

bagasse replied:
@timogoebel all is fine on your side now ?

timogoebel replied:
@bagasse: Yes. Thanks. Can you add a test for this to make sure it really works on all supported databases?

ares commented:
leaving the rest on @timogoebel but I think now when we have default value for MTU we should add validation to subnet that the mtu is required attribute (it might be caught by numerical validation already , I have not tested) but having presence validator I think will make sure the UI will add `*` next to the field, otherwise all my commets were resolved

bagasse commented:
@ares, numericaly validator imply presence by default (to allow blank/nil value, you have to provide :allow_nil option). Anyway I've added some test around this and default MTU value.

@timogoebel, added a test with the maximum MTU value.

timogoebel commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,22 @@ def vlanid
+
+      return self.subnet.mtu if self.subnet && self.subnet.mtu.present?
+      return self.subnet6.mtu if self.subnet6 && self.subnet6.mtu.present?
+      return nil
```
Nitpick: superfluous `return`

timogoebel commented on app/views/subnets/_fields.html.erb:
```diff
@@ -22,6 +22,7 @@
   <%= text_f f, :to,   :help_inline => _("Optional: Ending IP Address for IP auto suggestion") %>
 </div>
 <%= number_f f, :vlanid, :min => 0, :max => 4095, :help_inline => _("Optional: VLAN ID for this subnet") %>
+<%= number_f f, :mtu, :min => 68, :max => 4294967295, :help_inline => _("Optional: MTU for this subnet") %>
```
Is it really optional? I think this is kind of required as the database has a default value.

bagasse replied:
Done.

timogoebel commented on test/models/subnet/ipv6_test.rb:
```diff
@@ -85,6 +87,12 @@ class Subnet::Ipv6Test < ActiveSupport::TestCase
     assert_includes subnet.known_ips, '2001:db8::4'
   end
 
+  test "should not be able to save with MTU set to 4294967295" do
```
`... should be able to save ...`?

bagasse replied:
Done.

timogoebel commented on test/models/subnet/ipv6_test.rb:
```diff
@@ -85,6 +87,12 @@ class Subnet::Ipv6Test < ActiveSupport::TestCase
   end
 
+  test "should not be able to save with MTU set to 4294967295" do
+    subnet = FactoryBot.build(:subnet_ipv6)
```
FactoryBot.build(:subnet_ipv6, :mtu => 4294967295)

sean797 replied:
`FactoryBot.create` if you want to test database.

timogoebel commented on test/models/subnet/ipv6_test.rb:
```diff
@@ -85,6 +87,12 @@ class Subnet::Ipv6Test < ActiveSupport::TestCase
+  test "should not be able to save with MTU set to 4294967295" do
+    subnet = FactoryBot.build(:subnet_ipv6)
+    subnet.mtu = 4294967295
+    assert subnet.save
```
Maybe also check something like

```
assert_equal 4294967295, subnet.reload.mtu
```

bagasse replied:
done

timogoebel reviewed (changes requested):
@bagasse: Thanks. Looks good so far. I have some more comments.

ares commented:
> @ares, numericaly validator imply presence by default (to allow blank/nil value, you have to provide :allow_nil option). Anyway I've added some test around this and default MTU value.

yeah, that's what I thought, but our form helpers only check for explicit presence validator, so the form input does not have asterisk without `validates :mtu, :presence => true` being added to the subnet model. You can define it in subnet.rb and it would be inherited in both subnet types.

![mtu](https://user-images.githubusercontent.com/109773/33429148-e6db85dc-d5cb-11e7-942d-3a6d32a7cbc0.png)

bagasse commented:
@ares, thank for pointing that, I missed it. Added explicit presence validator.

timogoebel commented on app/views/subnets/_fields.html.erb:
```diff
@@ -22,6 +22,7 @@
   <%= text_f f, :to,   :help_inline => _("Optional: Ending IP Address for IP auto suggestion") %>
 </div>
 <%= number_f f, :vlanid, :min => 0, :max => 4095, :help_inline => _("Optional: VLAN ID for this subnet") %>
+<%= number_f f, :mtu, :min => 68, :max => 4294967295, :help_inline => _("MTU for this subnet") %>
```
Can we capitalize MTU?
`, :label => _('MTU')`

timogoebel commented on test/models/subnet_test.rb:
```diff
@@ -128,4 +128,15 @@ class Subnet::Test < Subnet; end
                  smart_variables)
   end
+
+  test "should not save if MTU is not provided" do
```
I think, we can replace this check with

```ruby
should validate_presence_of(:mtu)
```

to reduce the boilerplate.

timogoebel commented on test/models/subnet_test.rb:
```diff
@@ -128,4 +128,15 @@ class Subnet::Test < Subnet; end
+  end
+
+  test "should have MTU set to 1500 by default" do
+    assert_equal Subnet.new.mtu, 1500
```
please swap the arguments, `assert_equal expected, actual`

Should be `assert_equal 1500, Subnet.new.mtu`

timogoebel commented on test/models/subnet/ipv6_test.rb:
```diff
@@ -85,6 +87,12 @@ class Subnet::Ipv6Test < ActiveSupport::TestCase
+  test "should be able to save with MTU set to 4294967295" do
+    subnet = FactoryBot.build(:subnet_ipv6, :mtu => 4294967295)
+    assert subnet.save
+    assert_equal subnet.reload.mtu, 4294967295
```
Expected value (4294967295) should be the first parameter.

timogoebel reviewed (changes requested):
@bagasse: Thanks. Almost there. Another round of comments. :-)

bagasse commented:
@timogoebel, thanks ! Another round of fixes :)

timogoebel commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,22 @@ def vlanid
 
+    def mtu
+      # Determine a mtu according to the following cascading rules:
+      # 1. if the interface has a v4 subnet with a non-blank mtu, use that
```
As MTU is now a required attribute, this condition will never be false. So basically, this always prefers the IPv4 MTU. How about always taking the maximum or minimum value?

```ruby
[self.subnet.mtu, self.subnet6.mtu].select(&:present?).compact.max
```

The longer I think about that, I'm wondering if this is the right approach. The MTU is basically a setting on the ethernet layer. It defines the maximum size of the Ethernet frames. The subnets are on a layer on top the ethernet layer in the OSI model. Subnets are an IP thing.
I have a feeling, what we really need is to put this field on the nic model. We can define a default MTU for new hosts on the subnet. Better would be a dedicated "network" model that represents the ethernet layer or VLAN.

@ares, @sean797: What do you guys think?

ares replied:
In general, thumbs up. But that would be quite a change to vlan interfaces if we define vlan through network. Also would it be required for every user to define a network, or only in case they want to customize mtu/vlan on one place for all interfaces?

timogoebel reviewed (changes requested):
@bagasse: Thanks. This looks good, at least in theory. I do have some reservations regarding this left. Let's talk about them and then we're hopefully good to go.

Anyway, one more thing is missing: The MTU is not included in the APIv2 response. Could you extend the template, please? Adding `:mtu` to https://github.com/theforeman/foreman/blob/56562f3fd2ae494294456aee00c1d71b87106b0c/app/views/api/v2/subnets/main.json.rabl#L6 and maybe https://github.com/theforeman/foreman/blob/56562f3fd2ae494294456aee00c1d71b87106b0c/app/views/api/v2/interfaces/base.json.rabl#L3 should be enough.

ohadlevy commented:
On Dec 1, 2017 4:25 AM, "Timo Goebel" <notifications@github.com> wrote:

*@timogoebel* requested changes on this pull request.

@bagasse <https://github.com/bagasse>: Thanks. This looks good, at least in
theory. I do have some reservations regarding this left. Let's talk about
them and then we're hopefully good to go.

Anyway, one more thing is missing: The MTU is not included in the APIv2
response. Could you extend the template, please? Adding :mtu to
https://github.com/theforeman/foreman/blob/56562f3fd2ae494294456aee00c1d7
1b87106b0c/app/views/api/v2/subnets/main.json.rabl#L6 and maybe
https://github.com/theforeman/foreman/blob/56562f3fd2ae494294456aee00c1d7
1b87106b0c/app/views/api/v2/interfaces/base.json.rabl#L3 should be enough.
------------------------------

In app/models/nic/interface.rb
<https://github.com/theforeman/foreman/pull/5017#discussion_r154299639>:

> @@ -44,6 +44,22 @@ def vlanid
       return ''
     end

+    def mtu
+      # Determine a mtu according to the following cascading rules:
+      # 1. if the interface has a v4 subnet with a non-blank mtu, use that

As MTU is now a required attribute, this condition will never be false. So
basically, this always prefers the IPv4 MTU. How about always taking the
maximum or minimum value?

[self.subnet.mtu, self.subnet6.mtu].select(&:present?).compact.max

The longer I think about that, I'm wondering if this is the right approach.
The MTU is basically a setting on the ethernet layer. It defines the
maximum size of the Ethernet frames. The subnets are on a layer on top the
ethernet layer in the OSI model. Subnets are an IP thing.
I have a feeling, what we really need is to put this field on the nic
model. We can define a default MTU for new hosts on the subnet. Better
would be a dedicated "network" model that represents the ethernet layer or
VLAN.

@ares <https://github.com/ares>, @sean797 <https://github.com/sean797>:
What do you guys think?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/theforeman/foreman/pull/5017#pullrequestreview-80441355>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AABOxxCwR-cb9PlxWyHZPt3EDnt41yoRks5s78XwgaJpZM4Qlyu->
.


Does it make sense to split between logical networks (subnets) and
"physical" keeping vlan, mtu etc part of that new object? I'm assuming this
will also allow us to create multiple subnets instead of forcing ipv4 and
v6 at the same time?

timogoebel commented:
> Does it make sense to split between logical networks (subnets) and
> "physical" keeping vlan, mtu etc part of that new object?

@ohadlevy: Yes. I think so. It would also help with selecting the correct "compute resource network", what @bagasse hit earlier. I'm very sorry, the PRs always lead to this conclusion.
Maybe we can think of an intermediate solution? Adding a new "network" model is not a trivial task I think.

bagasse commented:
@ohadlevy; @timogoebel, I do agree with you, there is something weird with subnet in foreman, because as you said it already provides some network settings (like VLAN), and also IPv4 OR IPv6. It's a little schizophrenic :-)

I think your proposal to add a network model makes sense (and yes, it will not be trivial task), but how to implement it ? In regard of the network stack, I think that all 4 following implementations are valid and can be achieved. Some are really not good ideas, and personally, I never saw other implementation than **D** (1 network is 0..1 IPv4 subnet and 0..1 IPv6 subnet with at least one of them and IPv4 and IPv6 networks are not used on another network), except on some network routers (Case **C**). Case **D** is the more simple implementation but I don't know if some other are widely used by foreman users.

![screenshot from 2017-12-04 16-08-09](https://user-images.githubusercontent.com/2952632/33559732-0baf144c-d90e-11e7-9c77-54c52509b130.png)

Because actualy the VLAN ID is on the subnet model, and if you remove from the picture the Network and ComputeResourceNertwork parts, think foreman is acting as:
* Case **D** for virtual interfaces, because it will use the first VLANID it found and ignore the other. 
* Case **A** for 'normal' interfaces, if users don't care about VLAN tag.

IMHO, I think that the case **D** is the good way to go if we decide to go that path. I think that will help to automatically select the compute resource network in a second time, and will simplify the host interface GUI.

Any comments/thoughts are welcome :-)

timogoebel commented:
> Any comments/thoughts are welcome :-)

@bagasse: Thanks for the graphics. I've also given this some thought. I believe that option D is probably the way to go. My question is: Do we block the road to migrating to the "network model" in some way by this PR?

ares commented:
If this PR adds mtu attribute to v2 subnet, we'd need to keep it there forever, probably propagating the value to linked network once we have that model. Although the same applies to vlanid anyway. So I think this PR should not be blocked.

I like D, but IIUC that means having the network defined becomes a must even for users that don't need to adjust mtu/vlan. Perhaps it could be solved by providing default network with nil vlanid and default mtu. I can imagine migration for existing users this way too. Though having network and subnet objects meaning something different is slightly confusing for me as  networking amateur :-)

timogoebel commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,22 @@ def vlanid
+      # user's database. There is no way of determining the real mtu, so we
+      # pick the v4 one unless it turns out to be blank.
+
+      return self.subnet.mtu if self.subnet && self.subnet.mtu.present?
```
As the subnet mtu is now a required parameter, we need to change this as the mtu is always present.

How about taking the maximum value from both subnets for now?

```ruby
[self.subnet.mtu, self.subnet6.mtu].select(&:present?).compact.max
```

bagasse replied:
@timogoebel I don't think this a good idea on a network point of view. Maybe the good way to go will be to follow that @ares suggested, adding validation to this. I think we have to enforce this validations: validates that MTU and VLAN ID are identical for subnet and subnet6 on the same nic and hostgroup. This will be mostly ready on my side today, but error messages don't appears out of the box, I have to take a look at this (maybe because select2 is used for subnets in nic and hostgroup forms ?). I will add it as a separate commit so it will be more easy to revert it if you are not OK with this.
I think it will be nice to add filtering on subnet/subnet6 selects to only show subnets compatibles  if one is already selected. If we go this path, I think it will be more easy to add a network layer in the future. As there is no validations at the moment on vlanid, this can break some users, if they don't use the case **D** for their network topology. I can add a rake task to reports if there is some inconsistencies the the database.

Any comment ?

timogoebel replied:
> MTU and VLAN ID are identical for subnet and subnet6 on the same nic and hostgroup. 

Makes sense. I like that.

> This will be mostly ready on my side today, but error messages don't appears out of the box, I have to take a look at this (maybe because select2 is used for subnets in nic and hostgroup forms ?)

select2 should not be an issue there I believe. How do you add the errors?





> I think it will be nice to add filtering on subnet/subnet6 selects to only show subnets compatibles if one is already selected

Yes. That would be nice.

> As there is no validations at the moment on vlanid, this can break some users, if they don't use the case D for their network topology.

Yes. The vlan id can be empty right now. To be honest: Digitalocan, AWS, etc don't have the concept of a VLANID. I believe, we need to keep it optional.

timogoebel reviewed (changes requested):
@bagasse: Thanks for your patience. I think if we change the way the mtu is determined for an interface, this is ready to 🚢 .

Commit:
fixes #21694 - Add consistency validation for MTU and VLAN

timogoebel commented on app/validators/subnets_consistency_validator.rb:
```diff
@@ -0,0 +1,15 @@
+class SubnetsConsistencyValidator < ActiveModel::Validator
+  def validate(record)
+    if record.subnet && record.subnet6
```
Can be written with a guard clause, e.g.
`return unless record.subnet && record.subnet6`

timogoebel commented on test/models/hostgroup_test.rb:
```diff
@@ -389,6 +389,45 @@ class HostgroupTest < ActiveSupport::TestCase
   end
 
+  test 'with both subnet and subnet6 should be valid if VLAN ID is consistent between subnets' do
+    existing_domain = Domain.first
```
Please use `FactoryBot.create(:domain)`

timogoebel commented on test/models/hostgroup_test.rb:
```diff
@@ -389,6 +389,45 @@ class HostgroupTest < ActiveSupport::TestCase
+    subnet = FactoryBot.create(:subnet_ipv4, :domains => [existing_domain], :vlanid => 14)
+    subnet6 = FactoryBot.create(:subnet_ipv6, :domains => [existing_domain], :vlanid => 14)
+    hostgroup = FactoryBot.build(:hostgroup, :subnet => subnet, :subnet6 => subnet6)
+    assert hostgroup.valid?
```
we have a nice helper for this that provides nicer debugging output if the model is not valid.

```ruby
assert_valid hostgroup
```

timogoebel commented on app/validators/subnets_consistency_validator.rb:
```diff
@@ -0,0 +1,15 @@
+  def validate(record)
+    if record.subnet && record.subnet6
+      if record.subnet.vlanid != record.subnet6.vlanid
+        record.errors[:subnet_id] << _("VLAN ID is not consistent accross subnets")
```
Please prefer
```
record.errors.add(:subnet_id, _('"VLAN ID is not consistent accross subnets"'))
```

timogoebel reviewed (changes requested):
@bagasse: Thanks. Another round of feedback :-)

bagasse commented:
* Added consistency validation on host nics and hostgroups for vlanid and mtu.
* Filter to display only subnet compatible with the first selected to come.

bagasse commented:
Rebased. I've tried to add filters on subnets, but didn't find my way on the javascript part... Can we go head without it ?

timogoebel commented on app/validators/subnets_consistency_validator.rb:
```diff
@@ -0,0 +1,22 @@
+class SubnetsConsistencyValidator < ActiveModel::Validator
+  def validate(record)
+    return unless  record.subnet && record.subnet6
```
Nitpick: One space too much, between `unless` and `record.subnet`

timogoebel commented on test/fixtures/subnets.yml:
```diff
@@ -28,12 +29,13 @@ three:
-  mask: 255.255.255.0
-  tftp: two
-  vlanid: 44
+   name: four
```
Nitpick: indentation is off, YAML expects two spaces.

mmoll commented:
[test upgrade]
[test foreman]

bagasse commented:
Errors are related to rails migration, I've modified database migration file and wait for https://github.com/theforeman/foreman/pull/5133.

timogoebel reviewed (changes requested):
@bagasse: Thanks. I'm pretty busy right now, so only a preliminary review. I'll pick this up soon.

mmoll commented:
@bagasse Please update the respective line numbers in `config/as_deprecation_whitelist.yaml`, then the tests will run just fine.

bagasse commented:
[test upgrade]
[test foreman]

mmoll commented:
[test upgrade]

timogoebel commented on app/validators/subnets_consistency_validator.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def validate_vlanid(record)
+    return unless record.subnet.vlanid != record.subnet6.vlanid
```
`Vlanid` is currently not required. Should we ignore this if it's any one or both of the subnets?

bagasse replied:
This will be fine if both subnets don't have vlanid. IMHO having subnets with different vlanid (different values or one tagged and the other one not), is a nonsense.

timogoebel commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,22 @@ def vlanid
       return ''
     end
 
+    def mtu
```
I know, you must probably hate me right now, but I'm not too fond of this approach. This is based on my believe, that a host can have a different mtu than the mtu that is configured on the subnet.
Let me give you an example of what is not possible by this:
Let's say, I use jumbo frames and have configured the subnet for an mtu of 9000 bytes. One of my hosts is not configured properly and only has a mtu of 1500 bytes configured. As I user I'd want to see this in foreman. E.g. the host should upload facts to foreman and foreman should display the current value on the host page. Even better would be to display a warning (e.g. by a host status).

What do you think about adding a field "mtu" to the nic model? We could do it in another PR to get the subnet changes in faster.

bagasse replied:
Yes, I think you are right, we have to add this on per interface basis. I intended to add it in another PR, because I didn't have the time to dig into this. Do you want to add an mtu attribute to nic or get this from facts (or get it from facts and fallback to mtu attribute on nic ?) ? IMHO, I think that get it from facts and fallback to subnet value if not present is a good way to go.

I can add MTU to the show host's NICs tab too.

timogoebel replied:
> IMHO, I think that get it from facts and fallback to subnet value if not present is a good way to go.

I concur. I'd suggest: Let's remove this function (and dependent code) for now so we can get the first PR merged. Then let's add the code for adding mtu to an Interface in a separate PR. Deal?

timogoebel replied:
Never mind me, I was to slow. :-)

timogoebel reviewed (changes requested):
@bagasse: Thanks for your work on this. We definitely need to improve our networking model a lot, your help is very appreciated. Sorry for the long roundtrips in reviewing this, I'm afraid those small changes always have large consequences and I'm struggling to not block any improvements that may come in the future.

I left some more comments inline. In general: Works well, good job. :-)

bagasse commented:
@timogoebel, I've added the nic override part if mtu attr is present. I've also added a new line in the NICs tab on "show host" page.

timogoebel commented on app/models/nic/interface.rb:
```diff
@@ -44,6 +44,23 @@ def vlanid
+      # user's database. There is no way of determining the real mtu, so we
+      # pick the v4 one unless it turns out to be blank.
+
+      return self.attrs['mtu'] if !self.attrs['mtu'].blank?
```
Nit: Please prefer a positive statement: `if attrs['mtu'].present?`

timogoebel reviewed (changes requested):
@bagasse: I have one small remark, otherwise: Looks good. Thanks.
I still need to test this more thoroughly, but I have a feeling that we're almost there.

bagasse commented:
[test foreman]

mmoll commented:
@bagasse That one failure, although only occurring on MySQL, does look pretty related. ;)

bagasse commented:
@mmoll, I don't know why this was not falling on all databases, but anyway it's fixed ;)
@timogoebel, do you have any through/comment ? If all is OK I will squash all these commits (if needed).

timogoebel reviewed (dismissed):
Code looks good to me, thanks @bagasse.
I still need to test this a little, then we're ready to 🚢.

sean797 commented:
@bagasse can you rebase this please? I'll try and test it in the next few days.

bagasse commented:
@sean797, rebase done. Thank you for taking a look at this :)

mmoll commented:
@sean797 @timogoebel any news here? :point_up:

timogoebel commented on test/controllers/api/v2/subnets_controller_test.rb:
```diff
@@ -74,6 +74,7 @@ class Api::V2::SubnetsControllerTest < ActionController::TestCase
   test "delete destroys subnet not in use" do
     subnet = Subnet.first
     subnet.hosts.clear
+    subnet.hostgroups.clear
```
Is this needed? Why?

timogoebel reviewed (commented):
@bagasse: Sorry for the very emberrasing long turnaround on this.
I have one final question, then we're ready to go. I have successfully tested this. The code looks good. Nothing more to add from my side.

xprazak2 commented:
@sean797 @timogoebel, any additional comments?

sean797 reviewed (approved):
Sorry I thought I replied to this - tested the other week and works! :fireworks:

timogoebel reviewed (approved):
Never mind my question. I'll merge this so this can get into 1.18.
Thanks, @bagasse. Not just for the patch, but also for your patience.

bagasse commented:
@timogoebel, @sean797, @ares, @mmoll thank you for your advises and reviews !
